torch/nn/parallel/parallel_apply.py

```diff
@@ -16,12 +20,18 @@ def parallel_apply(modules, inputs, kwargs_tup=None):
     lock = threading.Lock()
     results = {}
 
+    def _get_device(obj):
+        if isinstance(obj, Variable):
+            return torch.cuda.device_of(obj)
+        if isinstance(obj, collections.Iterable):
+            vals = obj.values() if isinstance(obj, collections.Mapping) else obj
+            return next(dev for dev in map(_get_device, vals) if dev is not None)
```

Comment:
All sorts of other errors occur before here if there are no inputs :) These are all easily fixed, but there's a design choice to be made: if there are no inputs do zero or all devices run?
I think the latter is reasonable since it allows generating the input for each replica directly on the device (instead of broadcasting it). It also helps with failing fast since, instead of silently ignoring a bad null input, eventually the replicated module.